src/v2/dry_run.rs

```diff
@@ -0,0 +1,1063 @@
+}
+
+impl TryFrom<Option<Result<generated::DryRunResponse, tonic::Status>>>
+    for WithRemainingQuota<TransactionExecuted>
+{
+    type Error = DryRunError;
+
+    fn try_from(
+        value: Option<Result<generated::DryRunResponse, tonic::Status>>,
+    ) -> Result<Self, Self::Error> {
+        let response =
+            value.ok_or_else(|| tonic::Status::cancelled("server closed dry run stream"))??;
+        let quota_remaining = response.quota_remaining.require()?.into();
+        use generated::dry_run_response::*;
+        match response.response.require()? {
+            Response::Error(e) => {
+                let result = e.error.require()?.try_into()?;
+                if !matches!(
+                    result,
+                    ErrorResult::NoState()
+                        | ErrorResult::AccountNotFound()
+                        | ErrorResult::BalanceInsufficient {
+                            required_amount:  _,
+                            available_amount: _,
+                        }
+                        | ErrorResult::EnergyInsufficient { energy_required: _ }
+                ) {
+                    Err(tonic::Status::unknown("unexpected error response type"))?
+                }
+                Err(DryRunError::OperationFailed {
+                    result,
+                    quota_remaining,
+                })
+            }
+            Response::Success(s) => {
+                let response = s.response.require()?;
+                use generated::dry_run_success_response::*;
+                match response {
+                    Response::TransactionExecuted(res) => {
+                        let inner = TransactionExecuted {
+                            energy_cost:  res.energy_cost.require()?.into(),
+                            details:      res.details.require()?.try_into()?,
+                            return_value: res.return_value,
+                        };
+                        Ok(WithRemainingQuota {
+                            inner,
+                            quota_remaining,
+                        })
+                    }
+                    _ => Err(tonic::Status::unknown("unexpected success response type"))?,
+                }
+            }
+        }
+    }
+}
+
+type DryRunResult<T> = Result<WithRemainingQuota<T>, DryRunError>;
+
+/// A dry-run session.
+pub struct DryRun {
+    /// The channel used for sending requests to the server.
+    /// This is `None` if the session has been closed.
+    request_send:  Option<channel::mpsc::Sender<generated::DryRunRequest>>,
+    /// The channel used for receiving responses from the server.
+    response_recv: shared_receiver::SharedReceiver<
+        futures::stream::Fuse<tonic::Streaming<generated::DryRunResponse>>,
+    >,
+    /// The timeout in milliseconds for the dry-run session to complete.
+    timeout:       Option<u64>,
+    /// The energy quota for the dry-run session as a whole.
+    energy_quota:  Option<u64>,
```

Comment:
I think we should require the node to return this, and fail initial setup if it does not.